### PR TITLE
Update stats for Read and ReadAsync in random_access_file_reader for async prefetching

### PR DESCRIPTION
**Pull request by akankshamahajan15**
Summary: Update stats in random_access_file_reader for Read and
ReadAsync API to take into account the read latency for async
prefetching.

It also fixes ERROR_HANDLER_AUTORESUME_RETRY_COUNT stat whose value was
incorrect in portal.h

Test Plan: Update unit test

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: file/random_access_file_reader.h [hunk 1]
```diff
@@ -102,7 +110,8 @@ class RandomAccessFileReader {
       RateLimiter* rate_limiter = nullptr,
       const std::vector<std::shared_ptr<EventListener>>& listeners = {},
       Temperature file_temperature = Temperature::kUnknown,
-      bool is_last_level = false)
+      bool is_last_level = false, uint32_t async_hist_type = 0,
```
Comment (anand1976) [inline file/random_access_file_reader.h:113]: I don't know if its necessary to pass these. Are there scenarios where we'd want to use different histogram types? If not, we can simplify by hardcoding it. We do that for ticker stats anyway.

File: file/random_access_file_reader.cc [hunk 2]
```diff
@@ -425,19 +425,48 @@ IOStatus RandomAccessFileReader::PrepareIOOptions(const ReadOptions& ro,
...
+  return file_->ReadAsync(req, opts, read_async_callback, read_async_info,
+                          io_handle, del_fn, nullptr /*dbg*/);
+}
+
+void RandomAccessFileReader::ReadAsyncCallback(const FSReadRequest& req,
+                                               void* cb_arg) {
+  ReadAsyncInfo* read_async_info = static_cast<ReadAsyncInfo*>(cb_arg);
+  assert(read_async_info);
+  assert(read_async_info->cb);
+  if (stats_ != nullptr && file_async_read_hist_ != nullptr) {
```
Comment (anand1976) [inline file/random_access_file_reader.cc:466]: We also need to call RecordIOStats() and NotifyOnFileReadFinish().

File: file/random_access_file_reader.cc [hunk 3]
```diff
@@ -425,19 +425,48 @@ IOStatus RandomAccessFileReader::PrepareIOOptions(const ReadOptions& ro,
...
+  // Create a callback and populate info.
+  auto read_async_callback =
+      std::bind(&RandomAccessFileReader::ReadAsyncCallback, this,
+                std::placeholders::_1, std::placeholders::_2);
+  ReadAsyncInfo* read_async_info = new ReadAsyncInfo;
+  read_async_info->cb = cb;
+  read_async_info->cb_arg = cb_arg;
+
+  StopWatch sw(clock_, stats_, async_hist_type_,
+               (stats_ != nullptr) ? &(read_async_info->elapsed) : nullptr,
```
Comment (anand1976) [inline file/random_access_file_reader.cc:454]: Shouldn't we record the latency in ReadAsyncCallback?

File: file/random_access_file_reader.cc [hunk 4]
```diff
@@ -425,19 +425,71 @@ IOStatus RandomAccessFileReader::PrepareIOOptions(const ReadOptions& ro,
   }
 }
 
-// TODO akanksha: Add perf_times etc.
+// TODO akanksha:
+// 1. Add perf_level, notify listeners, iostats etc.
```
Comment (akankshamahajan15) [inline file/random_access_file_reader.cc:429]: Remove this line.

**Review by anand1976 — APPROVED** [review]
LGTM. Thanks for the quick turnaround!

**Pull request closed without merging**